### PR TITLE
Add support for Azure CDN endpoint URL

### LINKED ISSUE #84: Add option for artifacts to be available through Azure Content Delivery Network (CDN) URL instead of through Azure Storage Account blob URL
### What feature do you want to see added?

When the Azure Artifact Manager plugin is used to archive artifacts into an Azure Storage Account, those artifacts available through Jenkins from the Azure Storage Account are only available through the Azure Storage Account blob URL.

Because those artifacts are only available through the Azure Storage Account blob URL, the user clicking the link is downloading the artifact from the region that the Azure Storage Account is located in. Therefore, if those users are geographically distributed, they will potentially be downloading artifacts from a region far from their own and will have slower download speeds.

To mitigate this issue, the Azure Artifact Manager plugin can be updated to optionally support an Azure Content Delivery Network (CDN) URL in place of the Azure Storage Account blob URL. This will allow the Jenkins maintainer to optionally setup an Azure CDN endpoint URL so that when a user downloads an artifact from Jenkins, it will use the CDN URL instead which will pull the artifact from a region closer to the user, instead of from the region that the Azure Storage Account is located in.

### Upstream changes

Because the Azure Artifact Manager plugin is dependent on the Azure Storage plugin and a different CDN can be setup for every Azure Storage Account, the Azure Storage plugin will need to be updated to support linking an Azure CDN endpoint URL for a specific Azure Storage Account.

### Are you interested in contributing this feature?

Yes, I am actively working on implementing this feature across the Azure Storage plugin and the Azure Artifact Manager plugin.

### PR DESCRIPTION
**Pull request by Minreaux**
Resolves #84

### Description

These changes add support for an Azure CDN endpoint URL into the Azure Artifact Manager plugin, which will pull the Azure CDN endpoint URL from the configured Azure Storage Account, if and only if the Azure CDN endpoint URL is configured, and then distribute artifacts in Jenkins using the Azure CDN URL instead of the Azure Storage Account blob URL.

Otherwise, it will use the Azure Storage Account blob URL like normal. This PR is dependent on the upstream changes for the Azure Storage plugin here: https://github.com/jenkinsci/azure-storage-plugin/pull/280

I also fixed some issues I discovered while working on this so that I could get the plugin building locally without issue and to ensure that the Azure Storage and Azure Artifact Manager plugins had the same Jenkins version as a minimum dependency:

- Updated Jenkins version and bom to 2.387
- Fixed some warnings with the pom.xml

### Testing done

This change was tested locally in VSCode configured for Java/Jenkins development. I updated the JUnit tests to account for this new feature. I tested this locally using the `mvn verify` command.

Since these changes are dependent on the upstream changes I've made to the Azure Storage plugin, I tested these changes in a local instance of Jenkins using the `mvn hpi:run` command using a local version of the Azure Storage plugin containing the changes from this PR: https://github.com/jenkinsci/azure-storage-plugin/pull/280.

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Comment by Minreaux** [comment]
Tagging maintainer for review @timja.

File: src/test/java/com/microsoft/jenkins/artifactmanager/IntegrationTests/IntegrationTest.java [hunk 1]
```diff
@@ -27,6 +27,7 @@ protected static class TestEnvironment {
         public final String azureStorageAccountKey2;
 
         public final String blobURL;
+        public final String CdnURL;
```
Comment (timja) [inline src/test/java/com/microsoft/jenkins/artifactmanager/IntegrationTests/IntegrationTest.java:30]:
```suggestion
        public final String cdnURL;
```

File: src/test/java/com/microsoft/jenkins/artifactmanager/IntegrationTests/IntegrationTest.java [hunk 2]
```diff
@@ -41,7 +42,8 @@ protected static class TestEnvironment {
             azureStorageAccountKey2 = TestEnvironment.loadFromEnv("AZURE_STORAGE_TEST_STORAGE_ACCOUNT_KEY2");
 
             blobURL = Utils.DEF_BLOB_URL;
-            sampleStorageAccount = new StorageAccountInfo(azureStorageAccountName, azureStorageAccountKey1, blobURL);
+            CdnURL = "";
+            sampleStorageAccount = new StorageAccountInfo(azureStorageAccountName, azureStorageAccountKey1, blobURL, CdnURL);
```
Comment (timja) [inline src/test/java/com/microsoft/jenkins/artifactmanager/IntegrationTests/IntegrationTest.java:46]:
```suggestion
            sampleStorageAccount = new StorageAccountInfo(azureStorageAccountName, azureStorageAccountKey1, blobURL, cdnURL);
```

**Review by timja — APPROVED** [review]
Thanks! Not tested manually but looks simple enough

File: src/test/java/com/microsoft/jenkins/artifactmanager/IntegrationTests/IntegrationTest.java [hunk 3]
```diff
@@ -27,6 +27,7 @@ protected static class TestEnvironment {
         public final String azureStorageAccountKey2;
 
         public final String blobURL;
+        public final String cdnURL;
```
Comment (timja) [inline src/test/java/com/microsoft/jenkins/artifactmanager/IntegrationTests/IntegrationTest.java:30]:
```suggestion
```

File: src/test/java/com/microsoft/jenkins/artifactmanager/IntegrationTests/IntegrationTest.java [hunk 4]
```diff
@@ -41,7 +42,8 @@ protected static class TestEnvironment {
             azureStorageAccountKey2 = TestEnvironment.loadFromEnv("AZURE_STORAGE_TEST_STORAGE_ACCOUNT_KEY2");
 
             blobURL = Utils.DEF_BLOB_URL;
-            sampleStorageAccount = new StorageAccountInfo(azureStorageAccountName, azureStorageAccountKey1, blobURL);
+            CdnURL = "";
```
Comment (timja) [inline src/test/java/com/microsoft/jenkins/artifactmanager/IntegrationTests/IntegrationTest.java:45]:
```suggestion
```

File: src/test/java/com/microsoft/jenkins/artifactmanager/IntegrationTests/IntegrationTest.java [hunk 5]
```diff
@@ -41,7 +42,8 @@ protected static class TestEnvironment {
...
+            sampleStorageAccount = new StorageAccountInfo(azureStorageAccountName, azureStorageAccountKey1, blobURL, cdnURL);
```
Comment (timja) [inline src/test/java/com/microsoft/jenkins/artifactmanager/IntegrationTests/IntegrationTest.java:46]:
```suggestion
            sampleStorageAccount = new StorageAccountInfo(azureStorageAccountName, azureStorageAccountKey1, blobURL, "");
```

**Comment by Minreaux** [comment]
Thanks for all your help!

**Pull request merged**